File: backend/app/schemas/reports.py
from datetime import date
from decimal import Decimal

from pydantic import BaseModel, Field

from app.core.enums import StudentStatus


class ReportDateRange(BaseModel):
    start_date: date | None = None
    end_date: date | None = None


class ActiveStudentReportItem(BaseModel):
    id: int
    name: str
    cpf: str
    email: str
    status: StudentStatus


class ActiveStudentsReport(BaseModel):
    total: int = Field(ge=0)
    students: list[ActiveStudentReportItem]


class DefaulterStudentReportItem(BaseModel):
    student_id: int
    name: str
    cpf: str
    email: str
    overdue_amount: Decimal
    overdue_payments: int = Field(ge=0)


class DefaultersReport(BaseModel):
    total: int = Field(ge=0)
    students: list[DefaulterStudentReportItem]


class MostUsedPlanReportItem(BaseModel):
    plan_id: int
    plan_name: str
    enrollments_count: int = Field(ge=0)


class MostUsedPlansReport(BaseModel):
    plans: list[MostUsedPlanReportItem]


class RevenueSummaryReport(BaseModel):
    expected_revenue: Decimal
    received_revenue: Decimal
    overdue_revenue: Decimal
    pending_revenue: Decimal


class DailyAccessReportItem(BaseModel):
    date: date
    total_attempts: int = Field(ge=0)
    allowed_count: int = Field(ge=0)
    blocked_count: int = Field(ge=0)


class DailyAccessReport(BaseModel):
    days: list[DailyAccessReportItem]


class WorkoutSummaryReport(BaseModel):
    active_workout_plans: int = Field(ge=0)
    inactive_workout_plans: int = Field(ge=0)
    total_exercises: int = Field(ge=0)
    exercise_progress_records: int = Field(ge=0)

File: frontend/src/types/reports.ts
import type { Status } from "./common";
import type { FinancialStatus } from "./student";

export interface DateRangeParams {
  start_date?: string;
  end_date?: string;
}

export interface ActiveStudentReportItem {
  id: number;
  name: string;
  cpf: string;
  email: string;
  status: Status;
  financial_status: FinancialStatus;
}

export interface ActiveStudentsReport {
  total: number;
  students: ActiveStudentReportItem[];
}

export interface DefaulterStudentReportItem {
  student_id: number;
  name: string;
  cpf: string;
  email: string;
  overdue_amount: string;
  overdue_payments: number;
}

export interface DefaultersReport {
  total: number;
  students: DefaulterStudentReportItem[];
}

export interface MostUsedPlanReportItem {
  plan_id: number;
  plan_name: string;
  enrollments_count: number;
}

export interface MostUsedPlansReport {
  plans: MostUsedPlanReportItem[];
}

export interface RevenueSummaryReport {
  expected_revenue: string;
  received_revenue: string;
  overdue_revenue: string;
  pending_revenue: string;
}

export interface DailyAccessReportItem {
  date: string;
  total_attempts: number;
  allowed_count: number;
  blocked_count: number;
}

export interface DailyAccessReport {
  days: DailyAccessReportItem[];
}

export interface WorkoutSummaryReport {
  active_workout_plans: number;
  inactive_workout_plans: number;
  total_exercises: number;
  exercise_progress_records: number;
}
